semtypes/src/main/java/io/ballerina/types/TypeAtom.java

```diff
@@ -33,6 +38,21 @@ public static TypeAtom createTypeAtom(int index, AtomicType atomicType) {
 
     @Override
     public int hashCode() {
-        return index;
+        return this.getIdentifier().hashCode();
+    }
+
+    @Override
+    public Kind kind() {
+        if (atomicType instanceof ListAtomicType) {
```

Comment:
Would it be better if we implement a `kind()` method under the `AtomicType` interface and call that one here? Not sure if there will be much of a performance difference since we are replacing a chain of if else statements with an interface but I feel like it would be cleaner

Reply:
Fixed in [83a25ae](https://github.com/ballerina-platform/ballerina-lang/pull/42777/commits/83a25ae0038f4ad4203317ad58aa0ed3b6e4481e)